### PR TITLE
LightningModule.configure_optimizers: add type hints

### PR DESCRIPTION
**Pull request by adamjstewart**
## What does this PR do?

Adds type hints to `LightningModule.configure_optimizers`.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18463.org.readthedocs.build/en/18463/

### CODE REVIEW

File: src/lightning/pytorch/utilities/types.py [hunk 1]
```diff
@@ -84,6 +85,20 @@ class LRSchedulerConfig:
     strict: bool = True
 
 
+@dataclass
```
Comment (adamjstewart) [inline src/lightning/pytorch/utilities/types.py:88]: Could also use TypedDict

Comment (awaelchli) [reply src/lightning/pytorch/utilities/types.py:101]: In this format, the user returns a dictionary, not a dataclass. Same for the lr-scheduler config.

Comment (adamjstewart) [reply src/lightning/pytorch/utilities/types.py:101]: So you're saying I should use TypedDict for both LRSchedulerConfig and OptimizerLRSchedulerConfig?

**Comment by adamjstewart** [comment]
Btw, you'll likely be interested in https://github.com/pytorch/pytorch/pull/108464

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Pull request merged**